Adding Documentatoin for PX4 Ethernet setup (#1813)
PX4/PX4-user_guide: en/advanced_config/ethernet_setup.md

   ```

1. Copy and paste the following configuration information into the file (note: the indentations are important!):
   ```
   network:
     version: 2
     renderer: NetworkManager
     ethernets:
         enp2s0:
             addresses:
                 - 192.168.0.1/24
             nameservers:
                 addresses: [192.168.0.1]
             routes:
                 - to: 192.168.0.1
                   via: 192.168.0.1
   ```
   
   Save and exit the editor.
1. Apply the *netplan* configuration by entering the following command into the Ubuntu terminal.
   ```
   sudo netplan apply
   ```
   
If you have already configured the Pixhawk 5X, at this point you could connect it to the computer via an Ethernet cable and it would be automatically deteced by QGroundControl.
  
### Companion Computer Ethernet Network Setup

The setup for a companion computer will depend on the companion computer operating system.
A Linux operating system may support netplan, in which case the instructions would be the same as above, but using a unique IP address.

<!-- Then link or continue on with MAVSDK setup - depending on what works when you try it -->

## MAVSDK Sanity check

> Since we have to listen to port 14550 as mentioned earlier, the only thing changes in MAVSDK is the port we connect to UDP. 

Reviewer: This might "work" but is not correct. You'll want to change the port and streamed mavlink messages as outlined in " PX4 Ethernet Port Configuration". The set of messages a GCS needs and an onboard computer are different.

### Example for MAVSDK-Python:

To give an idea about how this can be done, we make a change in a MAVSDK-Python example in its Github repo. You can extend this to your own application accordingly. Change [this line](https://github.com/mavlink/MAVSDK-Python/blob/3c2ad41273a42597499d5c4ec4ad90c6a494e691/examples/telemetry.py#L10) to the one below in your local script:
```

>     await drone.connect(system_address="udp://:14550")

Reviewer: ```suggestion
await drone.connect(system_address="udp://:14550")
```

> ```
>     await drone.connect(system_address="udp://:14550")
> ```

Reviewer: ```suggestion
```python
await drone.connect(system_address="udp://:14550")
```
```

Successfully connected one gives you Telemetry information from your flight controller. 

### Example for ROS2:

Prerequisites:

> - You have a supported autopilot hardware with RTPS feature enabled firmware on it by using [this guide](https://docs.px4.io/master/en/middleware/micrortps.html#client-px4-px4-autopilot).

Reviewer: Use internal links
```suggestion
- You have a supported autopilot hardware with RTPS feature enabled firmware on it by using [this guide](../middleware/micrortps.md#client-px4-px4-autopilot).
```

- [ROS2](https://docs.px4.io/master/en/ros/ros2_comm.html#sanity-check-the-installation) has been set up correctly and [sanity check](https://docs.px4.io/master/en/ros/ros2_comm.html#sanity-check-the-installation) has been confirmed. 
- You have followed the Ethernet setup as of the top of this page. 

In this example it is assumed that you have followed the example to set your IP addresses.

1. Connect your Flight controller via Ethernet
2. Open **QGroundcontrol > Analyze Tools > MAVLink Console**
3. Enter the command below to start micro_rtps client on your flight controller. Note that the remote IP here is your companion computer IP. This by default starts the micrortps_client connected to UDP ports 2019 and 2020. To make changes you can take a look at [RTPS guide](https://docs.px4.io/master/en/middleware/micrortps.html#client-px4-px4-autopilot)

> ```

Reviewer: If you want stuff to be under a list, it needs to be at the correct indentation level - 3 spaces.

# PX4 Ethernet Setup

> Ethernet connectivity a faster, more reliable, and more flexible communication technology than serial communication approach.

Reviewer: Is this true? I mean 
-  is it really faster, more reliable and more flexible than a serial port? How much? 
- IF there is an Ethernet, is that what we recommend instead of USB for connecting QGC to PX4
- IF there is an Ethernet, is that what we recommend instead of USB or another serial cable for connecting a  companion to PX4?

I'm not doubting you - I just don't know. My guess is that it probably the main benefits are if you are connecting to IP radios and anything that is "natively IP".

Reviewer: Yeah, I'm not sure without "evidence". I think the main use case is likely to be to connect to things like IP radios.

I've adjusted the comment to make it clear that it is a great alternative, but not to add the value judgement "you should use this if you can".

This method is preferred over serial communication in case of direct connection between Pixhawk and IP Radio telemetry, companion computers and Networks.

> It utilizes UDP connection for MAVLINK communication over the port 14550 with [MAVLINK2](https://mavlink.io/en/guide/mavlink_2.html). 

Reviewer: Are you saying "by default PX4 will connect to the Ethernet on your specified address and stream MAVLink on port 14550?

1. That is the port for connecting a GCS which indicates that it probably streams the GCS profile, and that by default it is set up for connecting to a GCS. 
2. It may be that it also streams MAVLink for the companion profile to 14540 - if it does that then we'd state the default setup. Does it? If it does not then if you want to connect to a companion you will need to configure this.
3. So how do you configure the port in PX4 - is it just a serial port like any other from the PX4 perspective?

Author: 1 and 2. It is broadcasting on 14550 port, I tested both with MAVSDK and also QGC. 
This commit says it. 
https://github.com/PX4/PX4-Autopilot/pull/18716/commits/70ca144c4d0257a9c7f2f5dda04b72d62a10b1b1
3. MAVLINK instance already starts with that port.

Reviewer: @farhangnaderi No, that says just "1"

```
param set-default MAV_2_MODE 0  # `0` implies GCS. This would have to be `2` for MAVSDK or ROS
param set-default MAV_2_UDP_PRT 14550   # This port is for GCS. It might work with MAVSDK, but it is **not** for MAVSDK.
```

It is clear how you would set up "2", but not clear how you would set up `1` and `2`. I'll ask.

Reviewer: Chasing this up in https://github.com/PX4/PX4-Autopilot/issues/19343

Note, the port setting above proves that PX4 just things Ethernet is another serial port. So we will need to update https://docs.px4.io/master/en/peripherals/mavlink_peripherals.html as well. The good news is that then we will be able to link to that for the docs here. Worth waiting on answer to the issue though.

> ## Supported Flight Controllers

Reviewer: FYI only, here I've used headings to give the document structure and highlight relevant information.

Reviewer: PS Note, also I refer to flight controllers generally here, not just Pixhawk or FMU 5X. It is likely that in future there will non FMU devices that still integrate with Ethernet.

Author: Thanks @hamishwillee . 
I changed to FCU.

> PX4 supports Ethernet connectivity on [Pixhawk 5X-standard](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-011%20Pixhawk%20Autopilot%20v5X%20Standard.pdf) flight controllers that have an Ethernet port.

Reviewer: FYI, I changed this to the PDF. If you link the standards, use the github repo :-)

Author: Oh, ok. I did not know that.

Reviewer: No worries. The repo is the approved version.

It may also be supported on other boards.

Supported flight controllers include:

 - [Holybro Pixhawk 5X](../flight_controller/pixhawk5x.md)

> ## Connecting PX4 and a Ubuntu Linux Computer

Reviewer: FYI only, here I'm being very clear on what we are showing. This is not PX4 and a companion, or PX4 and Windows Computer. This matters because you use the word "PC" to refer to the computer. To me and anyone in my generation that implies a Windows computer :-)

Author: OK. I still have a long way to be clear. :D

To connect PX4 and a Ubuntu Linux computer you need to configure them to run on the same IP network.
On PX4 you update this configuration in the network configuration file on the SD card (`/fs/microsd/net.cfg`).
On Ubuntu you can do this by defining a netplan configuration file: `/etc/netplan/01-network-manager-all.yaml`

> Below we set up the systems on an IP network with addresses `192.168.0.Xxx`, where PX4 has address `192.168.0.4` and the computer has address `192.168.0.1`.

Reviewer: You use the word "protocol" a lot. I think you need to be more precise - below I use the term IP address scheme -  Because what I **think** you're doing here is setting  up a local subnet with ranges `192.168.0.Xxx`. Is that right? 

However I'm a bit rusty on networks, so can you confirm that:
- `192.168.0.4` is the hard coded address we are giving the flight controller.
- If so, why are we assigning a DNS server to PX4?
- How do you decide what address range to assign - is this just based on your router?
- For the Ubuntu, I have guessed that "the computer has address `192.168.0.1`" but I don't understand netplan config files. It could be that "192.168.0.1/24" means that it can have any address from 1 to 24.
  The main point here is can you explain what that netplan does, so that perhaps we can hint to users how they configure it themselves.
  
Upshot, it looks like you set up a subnet and then manually assign IP addresses in the range for each member. No DHCP etc. Is that right?

Author: - yes we have made it to that custom IP with netman commands within the previous steps. However, this is hardly coded and optional which could be customized. We have just made an assumption for the users who have no idea about networking. 
- I have followed the PR for Nuttx PR for that. 
- It can be customized actually, but since users may not have experience with networking, I assumed a whole setup based on an example. 
- 24 first bits of addressing. examples are [here](https://netplan.io/examples/)

Reviewer: Great thanks! What I've done here is provide links to the relevant docs and make it clear that this is an example that works with the PX4 setup. We don't need to explain everything, we just need to make it easy for people to find out.
